demo/src/demo_agent.rs

```diff
@@ -0,0 +1,275 @@
+pub struct MessageData {
+    message: String,
+    recipient_verkey: String,
+    sender_verkey: String,
+}
+
+async fn decrypt_as_msg<T>(wallet: Arc<dyn BaseWallet>, didcomm_msg: &[u8]) -> (T, String)
+where
+    T: for<'de> Deserialize<'de>,
+{
+    let decrypted_msg = wallet.unpack_message(&didcomm_msg).await.unwrap();
+    let unpacked: MessageData = serde_json::from_slice(&decrypted_msg).unwrap();
+    let msg: T = serde_json::from_str(&unpacked.message).unwrap();
+    return (msg, unpacked.sender_verkey);
+}
+
+pub static COLOR_YELLOW: &str = "\x1b[93m";
+pub static COLOR_GREEN: &str = "\x1b[32m";
+
+impl DemoAgent {
+    pub async fn new(
+        name: String,
+        mediator_base_url: Url,
+        mpsc_registry: Arc<MpscRegistry<UserMessage>>,
+        log_color: String,
+    ) -> DemoAgent {
+        let endpoint_url = mediator_base_url.join(&format!("/send_user_message/{name}")).unwrap();
+        let receiver = mpsc_registry.subscribe(&name).await;
+        let agent_uuid = uuid::Uuid::new_v4().to_string();
+        let wallet_config = WalletConfig {
+            wallet_name: format!("demo_{name}_{agent_uuid}"),
+            wallet_key: "8dvfYSt5d1taSd6yJdpjq4emkwsPDDLYxkNFysFD2cZY".into(),
+            wallet_key_derivation: "RAW".into(),
+            wallet_type: None,
+            storage_config: None,
+            storage_credentials: None,
+            rekey: None,
+            rekey_derivation_method: None,
+        };
+        let handle = create_and_open_wallet(&wallet_config).await.unwrap();
+        let wallet = IndySdkWallet { wallet_handle: handle };
+        info!("Created agent {name} with mediator endpoint {endpoint_url}");
+        DemoAgent {
+            name,
+            receiver,
+            wallet: Arc::new(wallet),
+            endpoint_url,
+            log_color,
+        }
+    }
+
+    /*
+    Waits until a new message address for the agent is received by mediator. When a message addressed
+    for the respective agent is received, it attempts to decrypt the message and deserialize into
+    expected type T.
+     */
+    pub async fn wait_for_message<T>(&self) -> Result<T, String>
+    where
+        T: for<'de> Deserialize<'de>,
+        T: Debug,
+    {
+        info!("{}Agent[{}] is expecting a message", self.name, self.log_color);
+        let didcomm_msg = self
+            .receiver
+            .lock()
+            .await
+            .recv()
+            .await
+            .ok_or::<String>("Channel closed".into())?;
+        let (aries_msg, sender_key) = decrypt_as_msg::<T>(self.wallet.clone(), didcomm_msg.as_slice()).await;
+        info!(
+            "{}Agent[{}] has received a message sender verkey: {sender_key}",
+            self.name, self.log_color
+        );
+        Ok(aries_msg)
+    }
+
+    pub async fn prepare_invitation(&self) -> (Invitation, PairwiseInfo) {
+        let (faber_pw_did, faber_invite_key) = self.wallet.create_and_store_my_did(None, None).await.unwrap();
```

Comment:
nit; not sure if you were looking for it , but you could also call `PairwiseInfo::create(wallet)` to avoid construction of `PairwiseInfo` at the end

Reply:
I would like us to actually get rid of the whole `PairwiseInfo` thing, which why I am not using it, rather creating it as necessity at the point where it's needed.
The "did" portion of `PairwiseInfo` is generally useless